### PR TITLE
Implement undo and redo

### LINKED ISSUE #38: Implement undo/redo
Should work on global state (#36).

- [x] Research
- [x] Implement
- [x] Document

### PR DESCRIPTION
**Pull request by thomaskrause**
## Pull request type

 

Feature

## What is the current behavior?

The user can't undo any changes.

Issue Number: This fixes #38 

## What is the new behavior?

- Undo and redo is implemented using the graph notification mechanism
- Editors need to call `ProjectManager.addCheckpoint` when a consistent change set (like adding a token) is finished. Salt events will be sent for each little updated, but the user probably wants to undo a complete user interaction
- When a checkpoint is created or restored, an `ANNOTATION_CHANGED` event is sent. This allows all editors to react for consistent changes to the annotation graph without fearing to deal with an inconsistent Salt model
- A toolbar with the new/open/save and both the undo and redo buttons has been added

## Does this introduce a breaking change?

Yes, the old events for small incremental changes have been replaced with `ANNOTATION_CHANGED`.
Also, the unused and untested `ProjectManager.loadDocumentFrom` function has been removed.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added **or** the PR is neither a bug fix nor a feature
- [x] Docs have been reviewed and added / updated if needed **or** the PR is neither a bug fix nor a feature
- [x] The *Unreleased* section in CHANGELOG.md has been amended to reflect the changes in this PR if needed
- [x] Build (`mvn verify`) was run locally and any changes were pushed
- [x] The pull request is against the correct branch (`master` for bug fixes, `develop` for new functionality)
- [x] [Dependencies and citation templates](https://github.com/hexatomic/hexatomic/tree/develop/releng/templates) have been updated where necessary **or** there are no new dependencies

---

# Checklist for maintainers

## Releases

- Do **NOT** push the green merge button on GitHub.  
Follow the latest version of the [developer/maintainer docs](https://hexatomic.github.io/hexatomic/dev/) for making releases.
- Check that license and citation information for dependencies are complete (each dependency has a folder in `/THIRD-PARTY/`).

### CODE REVIEW

**Review by sdruskat — CHANGES_REQUESTED** [review]
Thanks! Looking forward to seeing this in the product!

Albeit, the undo/redo buttons don't seem to update nicely, you always have to activate another view (or rather, the project manager view) for the buttons to change. I'm afraid that this doesn't work well for the user, so should be changed.

1. Open corpus
2. Open doc in graph editor
3. Do something
4. `Undo` doesn't visually update, although change is made
5. Click in Project Manager
6. `Undo` updates.
7. etc ...

button-behaviour

ctrl-behaviour

**Comment by thomaskrause** [comment]
Fix the issue with the non-enabled toolbar buttons by sending the necessary event (see 13.5 in https://www.vogella.com/tutorials/EclipseRCP/article.html) manually in the `ProjectManager`.

File: bundles/org.corpus_tools.hexatomic.core/Application.e4xmi [hunk 1]
```diff
@@ -12,14 +12,28 @@
       </children>
     </children>
     <mainMenu xmi:id="_tiknoMsLEemUdItNeSusZw" elementId="org.corpus_tools.hexatomic.core.menu.0">
-      <children xsi:type="menu:Menu" xmi:id="_ufSukMsLEemUdItNeSusZw" elementId="org.corpus_tools.hexatomic.core.menu.file" label="File">
+      <children xsi:type="menu:Menu" xmi:id="_ufSukMsLEemUdItNeSusZw" elementId="org.corpus_tools.hexatomic.core.menu.file" label="File" mnemonics="f">
```
Comment (sdruskat) [inline bundles/org.corpus_tools.hexatomic.core/Application.e4xmi:15]:
```suggestion
      <children xsi:type="menu:Menu" xmi:id="_ufSukMsLEemUdItNeSusZw" elementId="org.corpus_tools.hexatomic.core.menu.file" label="&amp;File">
```

Fix mnemonics to work. I don't know if you have tried this at all, but the mnemonics field doesn't do anything AFAIK. <kbd>Alt</kbd> + <kbd>f</kbd> work only when adding an `&amp;` before the actual mnemonic in the menu label. Perhaps this is different on you system though?

File: bundles/org.corpus_tools.hexatomic.core/Application.e4xmi [hunk 2]
```diff
@@ -12,14 +12,28 @@
...
         <children xsi:type="menu:HandledMenuItem" xmi:id="_P-HZsEy8EeqppqV5eQIggg" elementId="org.corpus_tools.hexatomic.core.handledmenuitem.save_salt_project" label="Save Salt Project" iconURI="platform:/plugin/org.corpus_tools.hexatomic.core/icons/fontawesome/save-regular.png" mnemonics="" command="_o-7oQEy7EeqppqV5eQIggg"/>
         <children xsi:type="menu:HandledMenuItem" xmi:id="_xiM4wEzaEeqppqV5eQIggg" elementId="org.corpus_tools.hexatomic.core.handledmenuitem.save_as_salt_project" label="Save Salt Project As..." iconURI="" command="_mqJxIEzaEeqppqV5eQIggg"/>
         <children xsi:type="menu:HandledMenuItem" xmi:id="_ZKoIMFPxEeq55bpbFwb6CQ" elementId="org.corpus_tools.hexatomic.core.handledmenuitem.exit" label="Exit" iconURI="platform:/plugin/org.corpus_tools.hexatomic.core/icons/fontawesome/times-solid.png" command="_HVy0oFPxEeq55bpbFwb6CQ"/>
       </children>
+      <children xsi:type="menu:Menu" xmi:id="_c2kBYJ6CEeqcNd4BgS8ZlA" elementId="org.corpus_tools.hexatomic.core.menu.edit" label="Edit" mnemonics="e">
```
Comment (sdruskat) [inline bundles/org.corpus_tools.hexatomic.core/Application.e4xmi:22]:
```suggestion
      <children xsi:type="menu:Menu" xmi:id="_c2kBYJ6CEeqcNd4BgS8ZlA" elementId="org.corpus_tools.hexatomic.core.menu.edit" label="&amp;Edit">
```

Same as above, mnemonics only work the old school SWT way.

**Review by sdruskat — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing this.

Unfortunately I've stumbled over another bug, not sure if it was introduced in this PR, but should probably be fixed in this one: Saving throws an error.

1. Create new doc via "Add" buttons
2. Open in Grid Editor
3. Tokenize something
4. Save as project somewhere
5. Do another tokenization.
6a. Sometimes it works and new tokens are added
6b. Sometimes it doesn't, and then :boom: `InvocationTargetException` on Save

bug

```
16:16:06.701 [main] INFO  o.c.h.c.ApplicationLifecycle -
                Logging configured from internal configuration
16:16:06.732 [main] DEBUG o.c.h.c.ProjectManager -
                Starting Project Manager
16:16:19.625 [main] DEBUG o.c.h.c.ProjectManager -
                Adding 3 changes as checkpoint to undo list
16:16:20.016 [main] DEBUG o.c.h.c.ProjectManager -
                Adding 4 changes as checkpoint to undo list
16:16:20.370 [main] DEBUG o.c.h.c.ProjectManager -
                Adding 10 changes as checkpoint to undo list
16:16:20.371 [main] DEBUG o.c.h.c.CorpusStructureView -
                Selecting created document
16:16:24.872 [main] DEBUG o.c.h.c.ProjectManager -
                Adding 41 changes as checkpoint to undo list
16:16:24.873 [main] DEBUG o.c.h.g.GraphEditor -
                Received ANNOTATION_CHANGED event for changeset org.corpus_tools.hexatomic.core.undo.ChangeSet@1eca3ea7
16:16:42.155 [main] DEBUG o.c.h.c.ProjectManager -
                Adding 31 changes as checkpoint to undo list
16:16:42.156 [main] DEBUG o.c.h.g.GraphEditor -
                Received ANNOTATION_CHANGED event for changeset org.corpus_tools.hexatomic.core.undo.ChangeSet@3a3f96ab
16:16:51.830 [main] ERROR o.c.h.c.ProjectManager -
                Exception occured: Could not save project
java.lang.reflect.InvocationTargetException: null
	at org.eclipse.jface.operation.ModalContext.run(ModalContext.java:397) ~[na:na]
	at org.eclipse.jface.dialogs.ProgressMonitorDialog.run(ProgressMonitorDialog.java:469) ~[na:na]
	at org.corpus_tools.hexatomic.core.ProjectManager.saveTo(ProjectManager.java:405) ~[na:na]
	at org.corpus_tools.hexatomic.core.ProjectManager.save(ProjectManager.java:473) ~[na:na]
	at org.corpus_tools.hexatomic.core.handlers.SaveHandler.execute(SaveHandler.java:46) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:58) ~[na:na]
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:318) ~[na:na]
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:252) ~[na:na]
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:173) ~[na:na]
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:156) ~[na:na]
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:488) ~[na:na]
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:487) ~[na:na]
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:213) ~[na:na]
	at org.eclipse.e4.ui.workbench.renderers.swt.HandledContributionItem.executeItem(HandledContributionItem.java:438) ~[na:na]
	at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem.handleWidgetSelection(AbstractContributionItem.java:449) ~[na:na]
	at org.eclipse.e4.ui.workbench.renderers.swt.AbstractContributionItem.lambda$2(AbstractContributionItem.java:475) ~[na:na]
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89) ~[na:na]
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5687) ~[na:na]
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1423) ~[na:na]
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4955) ~[na:na]
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4448) ~[na:na]
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1160) ~[na:na]
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:338) ~[na:na]
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1049) ~[na:na]
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155) ~[na:na]
	at org.eclipse.e4.ui.internal.workbench.swt.E4Application.start(E4Application.java:166) ~[na:na]
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:203) ~[na:na]
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137) ~[na:na]
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107) ~[na:na]
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:401) ~[na:na]
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:657) ~[org.eclipse.equinox.launcher_1.5.700.v20200207-2156.jar:na]
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:594) ~[org.eclipse.equinox.launcher_1.5.700.v20200207-2156.jar:na]
	at org.eclipse.equinox.launcher.Main.run(Main.java:1447) ~[org.eclipse.equinox.launcher_1.5.700.v20200207-2156.jar:na]
	at org.eclipse.equinox.launcher.Main.main(Main.java:1420) ~[org.eclipse.equinox.launcher_1.5.700.v20200207-2156.jar:na]
Caused by: java.lang.NullPointerException: null
	at org.corpus_tools.salt.util.internal.persistence.SaltXML10Writer.writeRelation(SaltXML10Writer.java:646) ~[na:na]
	at org.corpus_tools.salt.util.internal.persistence.SaltXML10Writer.writeDocumentGraph(SaltXML10Writer.java:422) ~[na:na]
	at org.corpus_tools.salt.util.internal.persistence.SaltXML10Writer.writeDocumentGraph(SaltXML10Writer.java:343) ~[na:na]
	at org.corpus_tools.salt.util.SaltUtil.saveDocumentGraph(SaltUtil.java:609) ~[na:na]
	at org.corpus_tools.salt.common.impl.SDocumentImpl.saveDocumentGraph(SDocumentImpl.java:189) ~[na:na]
	at org.corpus_tools.hexatomic.core.ProjectManager$1.run(ProjectManager.java:343) ~[na:na]
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122) ~[na:na]
```

**Comment by thomaskrause** [comment]
> Unfortunately I've stumbled over another bug, not sure if it was introduced in this PR, but should probably be fixed in this one: Saving throws an error.

I checked and this bug is present in the current master branch (#220) and I would propose to fix this as a hotfix release (which should of course include a regression test). It might be related to #214 but this will only be clear after we find the root cause of this issue, which is obviously not the changes added in this PR.

**Comment by thomaskrause** [comment]
> Unfortunately I've stumbled over another bug, not sure if it was introduced in this PR, but should probably be fixed in this one: Saving throws an error.

This is now fixed in `master` and integrated into this PR

File: bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/ChangeSet.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.Optional;
+import java.util.Set;
+import org.corpus_tools.hexatomic.core.SaltHelper;
+import org.corpus_tools.salt.common.SDocumentGraph;
+
+public class ChangeSet {
+
+  private final List<ReversibleOperation> changes = new LinkedList<>();
+
+  public ChangeSet(Collection<ReversibleOperation> changes) {
```
Comment (sdruskat) [inline bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/ChangeSet.java:37]: Wondering why Checkstyle hasn't complained about missing JavaDoc here ...

File: bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/AddCorpusGraphOperation.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import org.corpus_tools.salt.common.SCorpusGraph;
+import org.corpus_tools.salt.common.SaltProject;
+
+public class AddCorpusGraphOperation implements ReversibleOperation {
+
+  private final SaltProject project;
+  private final SCorpusGraph corpusGraph;
+
+  /**
+   * Create a undo operation for adding a corpus graph to a Salt project.
```
Comment (sdruskat) [inline bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/AddCorpusGraphOperation.java:33]:
```suggestion
   * A reversible operation for adding a corpus graph to a Salt project.
```

File: bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/AddLayerToGraphOperation.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import org.corpus_tools.salt.graph.Graph;
+import org.corpus_tools.salt.graph.Layer;
+
+public class AddLayerToGraphOperation<L extends Layer<?, ?>> implements ReversibleOperation {
+
+  private final L layer;
+  private final Graph<?, ?, L> graph;
+
+  /**
+   * Create a undo operation for a layer that was added to a graph.
```
Comment (sdruskat) [inline bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/AddLayerToGraphOperation.java:34]:
```suggestion
   * A reversible operation for a layer that was added to a graph.
```

Using the verb IMHO is imprecise. Here and in the other classes implementing ReversibleOperation.

File: bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/AddNodeToGraphOperation.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import org.corpus_tools.salt.graph.Graph;
+import org.corpus_tools.salt.graph.Node;
+
+public class AddNodeToGraphOperation<N extends Node> implements ReversibleOperation {
+
+  private final N node;
+  private final Graph<N, ?, ?> graph;
+
+  /**
+   * Create a undo operation for a node that was added to a graph.
```
Comment (sdruskat) [inline bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/AddNodeToGraphOperation.java:34]:
```suggestion
   * A reversible operation for a node that was added to a graph.
```

File: bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/AddNodeToLayerOperation.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+import org.corpus_tools.salt.graph.Node;
+
+public class AddNodeToLayerOperation<N extends Node>
+    implements ReversibleOperation {
+
+  private final Layer<N, ?> layer;
+  private final N node;
+  
+  /**
+   * Create a reversible operation for adding a node to a layer.
```
Comment (sdruskat) [inline bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/AddNodeToLayerOperation.java:35]:
```suggestion
   * A reversible operation for adding a node to a layer.
```

File: bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/AddRelationToGraphOperation.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import org.corpus_tools.salt.graph.Relation;
+
+public class AddRelationToGraphOperation<N extends Node, R extends Relation<N, N>>
+    implements ReversibleOperation {
+
+  private final R rel;
+  private final Graph<N, R, ?> graph;
+
+  /**
+   * Create a undo operation for a relation that was added to a graph.
```
Comment (sdruskat) [inline bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/AddRelationToGraphOperation.java:36]:
```suggestion
   * A reversible operation for a relation that was added to a graph.
```

File: bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/LabelAddOperation.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+import org.corpus_tools.hexatomic.core.undo.ReversibleOperation;
+import org.corpus_tools.salt.graph.LabelableElement;
+
+public class LabelAddOperation implements ReversibleOperation {
+
+  private final String qname;
+
+  private final LabelableElement container;
+
+  public LabelAddOperation(LabelableElement container, String qname) {
```
Comment (sdruskat) [inline bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/LabelAddOperation.java:33]: Again, weird why Checkstyle didn't complain about missing Javadoc ...

File: docs/dev/src/development/add-new-editor.md [hunk 10]
```diff
@@ -78,6 +78,28 @@ String documentID = part.getPersistedState()
 Optional<SDocument> doc = projectManager.getDocument(documentID);
 ```
 
+See the [Data model](./data-model/#reacting-to-changes) section on how to listen to changes on the document.
+
+## Changing the document
+
+You can manipulate the document graph returned by the `ProjectManager` directly using the Salt API.
+Every change via the Salt API will automatically be recorded to enable undoing and redoing the changes.
+A user interaction that manipulates the document graph can consist of multiple Salt API calls and you must mark the end of a user interaction by calling the `addCheckpoint()` function of the injected `ProjectManager` instance.
```
Comment (sdruskat) [inline docs/dev/src/development/add-new-editor.md:87]:
```suggestion
A user interaction that manipulates the document graph can consist of multiple Salt API calls, each one of which may potentially lead to inconsistent states. You must therefore mark the end of a user interaction by calling the `addCheckpoint()` function of an injected `ProjectManager` instance.
```

I think this change suggestion better explains why you actually need to add checkpoints, but feel free to edit.

File: docs/dev/src/development/automated-tests/README.md [hunk 11]
```diff
@@ -25,6 +25,9 @@ To specifically run the tests and not install the artifacts, use `mvn integratio
 Unlike unit tests, `mvn test` will not work for integration tests, as the Tycho Surefire Plugin requires
 the bundles to be packaged, which happens in Maven's `package` phase, which comes after `test` and before `integration-test`.
 
+You can also generate a test code coverage report by executing `mvn verify site -Pcoverage`. 
+The generated report will be located under `tests/org.corpus_tools.hexatomic.tests.report/target/site/jacoco-aggregate/index.html`.
+
 To learn more about the Maven build lifecycle, read the [Maven Lifecycles Reference](http://web.archive.org/web/20191128092924/https://maven.apache.org/ref/3.6.2/maven-core/lifecycles.html).
 
 You can also generate a test code coverage report by executing `mvn verify site -Pcoverage`.
```
Comment (sdruskat) [inline docs/dev/src/development/automated-tests/README.md:33]:
```suggestion
```

Remove duplicated paragraph

File: docs/dev/src/development/data-model/README.md [hunk 12]
```diff
@@ -37,18 +37,23 @@ public class MyClass {
...
+## Reacting to changes
+
+You can register for any changes on the Salt project (e.g., added annotations to a document), by subscribing to the `Topics.ANNOTATION_CHANGED` topic, which will be sent by the [Eclipse RCP `IEventBroker` service](http://web.archive.org/web/20200427021644/https://www.vogella.com/tutorials/Eclipse4EventSystem/article.html).
+After registering your listener, you will receive *all* updates for *all* documents and *all* changes to the project structure.
+You have to decide in your own code if you need to handle an update, e.g., because it is related to a document you are editing for which the event should trigger redrawing the editor you are implementing.
+The argument for the `Topics.ANNOTATION_CHANGED` event is of the type `org.corpus_tools.hexatomic.core.undo.ChangeSet` and contains a list of all changes.
+It also has a helper function to test if a given document is affected by its changes.
```
Comment (sdruskat) [inline docs/dev/src/development/data-model/README.md:46]:
```suggestion
It also has a helper function to test if a given document is affected by its changes.

```

Add blank line for readability.

File: pom.xml [hunk 13]
```diff
@@ -75,8 +75,7 @@
 							</goals>
 							<phase>process-sources</phase>
 							<configuration>
-								<organizationName>Stephan Druskat,
-									Thomas Krause</organizationName>
+								<organizationName>Stephan Druskat, Thomas Krause</organizationName>
```
Comment (sdruskat) [inline pom.xml:78]: I've tried to fix the respective bit in the POM configuration for the license plugin, but this keeps coming back :angry:.

File: tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestCorpusStructure.java [hunk 14]
```diff
@@ -217,4 +217,36 @@ void testDelete() {
     bot.toolbarButton(DELETE_BUTTON_TEXT).click();
     assertTrue(project.getCorpusGraphs().isEmpty());
   }
+
+  @Test
+  @Order(3)
+  void testUndoRedo() {
```
Comment (sdruskat) [inline tests/org.corpus_tools.hexatomic.it.tests/src/main/java/org/corpus_tools/hexatomic/it/tests/TestCorpusStructure.java:223]:
```suggestion
  void testUndo() {
```

Doesn't test redos.

**Review by sdruskat — APPROVED** [review]
Approved with very few minor change suggestions.

How about that :tada:!

File: bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/ChangeSet.java (hunk 3 above)
Comment (thomaskrause) [reply bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/ChangeSet.java:37]: For simple constructors that just initialize a value with one argument seem to be handled like getter/setter functions which also don't need a JavaDoc.

File: bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/LabelAddOperation.java (hunk 9 above)
Comment (thomaskrause) [reply bundles/org.corpus_tools.hexatomic.core/src/main/java/org/corpus_tools/hexatomic/core/undo/operations/LabelAddOperation.java:33]: See explanation above

**Pull request merged**